### PR TITLE
Image ToTensor operator - GPU support, 3D/4D inputs

### PR DESCRIPTION
**Pull request by sandeep-krishnamurthy**
## Description ##

1. <s>Make to_tensor as operator. No contribution to gradient.</s>Add description, InPlaceOption and other parameters during operator registration.
2. Support CPU and GPU. This PR adds GPU supports and re-organizes previously supported CPU.
3. to_tensor transformation operator can take 3D (c, h, w) or 4D (n, c, h, w) at once. So that a batch of input can be passed in the transformation.
4. Parallelization with OMP
5. This is backward compatible change. No existing functionality will be affected.

# Background 
We can enable users to fuse transformation operators with network graph and export it. This end to end network with transforms + network will greatly simplify inference deployment.
Related PR for normalize transformation - https://github.com/apache/incubator-mxnet/pull/13802/

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [X] Changes are complete (i.e. I finished coding on this PR)
- [X] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- [X] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [X] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- Make to_tensor to accept 3D or 4D input
- CPU / GPU support
- No gradient contribution
- Reorganize like any other operators

## Comments ##
- A related PR was raised here https://github.com/apache/incubator-mxnet/pull/13614 reviewed by @apeforest and @stu1130 - I have addressed those comments. I am closing other PR as it was mixing multiple transforms and functionality.

@apeforest @stu1130 @zhreshold @nswamy - For review

### CODE REVIEW

File: src/operator/image/totensor_op-inl.h [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+ * Copyright (c) 2018 by Contributors
+ * \file totensor_op-inl.h
+ * \brief Image to tensor operator
+*/
+#ifndef MXNET_OPERATOR_IMAGE_TOTENSOR_OP_INL_H_
+#define MXNET_OPERATOR_IMAGE_TOTENSOR_OP_INL_H_
+
+
+#include <mxnet/base.h>
+#include <algorithm>
```
[the remark this answers is not in the document]
Comment (sandeep-krishnamurthy) [reply src/operator/image/totensor_op-inl.h]: Thanks for flagging fixed it.

**Comment by sandeep-krishnamurthy** [comment]
@stu1130 - Thanks, I have addressed your comments. Can you please take a look?
@zhreshold - Can you please help review this PR?

**Comment by sandeep-krishnamurthy** [comment]
@szha 
As commented in other PR - https://github.com/apache/incubator-mxnet/pull/13802
1. I did not mean it is not an operator. I meant in this PR we reorganize, cpu/gpu support, and more such properties of an operator are added. Reworded the description.
2. IMO, the reorganizing and creation of new files for this operator is in the right direction and unifies with other operator pattern. Having all these implementation in image_random-inl.h creates more confusion and out of order compared to other operators.

**Comment by szha** [comment]
From first glance there doesn't seem to be that many changes to the extent that it would cause confusion. Let's at least give it a try.

**Comment by zhreshold** [comment]
@sandeep-krishnamurthy Can you keep the ops in the original files for sake of easier review? For now I get lost what has actually been modifed in addtion to the original implementation.

If you find the name of the file not appropriate, you can file a separate PR to move it and notify me that the new PR is purely to re-org code.

**Comment by sandeep-krishnamurthy** [comment]
> @sandeep-krishnamurthy Can you keep the ops in the original files for sake of easier review? For now I get lost what has actually been modifed in addtion to the original implementation.
> 
> If you find the name of the file not appropriate, you can file a separate PR to move it and notify me that the new PR is purely to re-org code.

Yes. I was working on moving it back per @szha suggestion.

**Comment by sandeep-krishnamurthy** [comment]
@zhreshold , @szha - I moved back all changes to same files as per suggestion. Can you please take a look at this PR? Thanks!

File: src/operator/image/image_random.cc [hunk 2]
```diff
@@ -39,14 +39,43 @@ DMLC_REGISTER_PARAMETER(RandomLightingParam);
...
+              [ 0.96470588  0.1882353 ]]
+             [[ 0.6156863   0.73725492]
+              [ 0.46666667  0.98039216]
+              [ 0.44705883  0.45490196]
+              [ 0.01960784  0.8509804 ]]
+             [[ 0.39607844  0.03137255]
+              [ 0.72156864  0.52941179]
+              [ 0.16470589  0.7647059 ]
+              [ 0.05490196  0.70588237]]]
+             <NDArray 3x4x2 @cpu(0)>
```
Comment (zhreshold) [inline src/operator/image/image_random.cc]: need a 4D example

Comment (sandeep-krishnamurthy) [reply src/operator/image/image_random.cc]: Updated with an 4D example

**Comment by sandeep-krishnamurthy** [comment]
@szha - I have addressed your comments. Can you please take a look at this PR? Thanks.

**Pull request merged**